### PR TITLE
Map UI improvements

### PR DESCRIPTION
**Pull request by MartinaeyNL**
Work in progress.

### CODE REVIEW

File: ui/component/or-map/src/index.ts [hunk 1]
```diff
@@ -467,6 +495,38 @@ export class OrMap extends LitElement {
         this.addEventListener(OrMapMarkerChangedEvent.NAME, this._onMarkerChangedEvent);
     }
 
+    public addMarker(asset: Asset): boolean {
+        let res = asset.attributes?.hasOwnProperty("location");
+        if (res && asset.attributes && asset.attributes.location.value) {
+            let coordinates = asset.attributes.location.value;
+            this._map?.addMark(asset.id ? asset.id : '', asset.name ? asset.name : '', asset.type ? asset.type : '', coordinates.coordinates[0], coordinates.coordinates[1], asset);
```
Comment (Ekhorn) [inline ui/component/or-map/src/index.ts]:
```suggestion
            this._map?.addMark(asset.id ?? '', asset.name ?? '', asset.type ?? '', coordinates.coordinates[0], coordinates.coordinates[1], asset);
```

File: ui/component/or-map/src/mapwidget.ts [hunk 2]
```diff
@@ -656,6 +687,433 @@ export class MapWidget {
...
+                                        currentValue: assetAttributeValue
+                                    };
+
+                                    iconAndColour = getMarkerIconAndColorFromAssetType(assetType, overrideOpts);
+                                } else {
+                                    //console.log('no colours treshold defined on asset type ' + assetType);
+                                }
+
+
+                                let findIndex = props.findIndex((prop: any) => { return assetType === prop.assetType});
```
Comment (Ekhorn) [inline ui/component/or-map/src/mapwidget.ts]:
```suggestion
                                let findIndex = props.findIndex((prop: any) => assetType === prop.assetType);
```

File: ui/app/manager/src/pages/page-map.ts [hunk 3]
```diff
@@ -340,32 +386,35 @@ export class PageMap extends Page<MapStateKeyed> {
     }
 
     protected render() {
+        let assetIdsToShow: string[] = this._map ? this._map.getCurrentView() : [];
 
         return html`
             
             ${this._currentAsset ? html `<or-map-asset-card .config="${this.config?.card}" .assetId="${this._currentAsset.id}" .markerconfig="${this.config?.markers}"></or-map-asset-card>` : ``}
+
+            <or-map-legend .assetTypes="${this.assetTypes}"></or-map-legend>
```
Comment (Ekhorn) [inline ui/app/manager/src/pages/page-map.ts]: Show this conditionally when `assetTypes > 1`

Comment (Ekhorn) [reply ui/app/manager/src/pages/page-map.ts]: See https://github.com/openremote/openremote/issues/2201

File: ui/app/manager/src/pages/page-map.ts [hunk 4]
```diff
@@ -375,30 +424,48 @@ export class PageMap extends Page<MapStateKeyed> {
...
         router.navigate(getMapRoute());
     }
 
+    protected onMapMoved(e: OrMapMovedEvent) {
+        this.movedDone = new Date();
+    }
+
+    protected onMapLoad(e: OrMapSourceLoadedEvent) {
+        this.mapInitLoaded = !this.mapInitLoaded;
+        this._map.renderCurrentCluster(this.config.markers);
```
Comment (Ekhorn) [inline ui/app/manager/src/pages/page-map.ts]: Consider removing as https://github.com/openremote/openremote/commit/8053b512f5497de25a5f3e9af21cdf85ac74fa64 showed during testing that clustering remains functional while improving performance significantly.


```suggestion
```

File: ui/component/or-map/src/index.ts [hunk 5]
```diff
@@ -467,6 +495,38 @@ export class OrMap extends LitElement {
...
+        } else {
+            return [];
+        }
+    }
+
+    public renderCurrentCluster(config?: MapMarkerAssetConfig | undefined): void {
+        if (this._map) {
+            this._map.renderCurrentCluster(config);
+        }
+    }
```
Comment (Ekhorn) [inline ui/component/or-map/src/index.ts:395]: Please see https://github.com/openremote/openremote/pull/1392/files#r2444858335

```suggestion
```

File: ui/component/or-map/src/index.ts [hunk 6]
```diff
@@ -539,6 +599,8 @@ export class OrMap extends LitElement {
                 .setControls(this.controls)
                 .setGeoJson(this.geoJson);
             this._map.load().then(() => {
+                this.loadPoints();
+
```
Comment (Ekhorn) [inline ui/component/or-map/src/index.ts]: Consider removing as https://github.com/openremote/openremote/commit/e9afa998cf3b11b68ec16a36f13a40c87e338fe4 showed during testing that clustering remains functional while improving performance significantly.

```suggestion
```

File: ui/component/or-map/src/mapwidget.ts [hunk 7]
```diff
@@ -656,6 +687,433 @@ export class MapWidget {
...
+            r0,
+            0,
+            largeArc,
+            0,
+            r + r0 * x0,
+            r + r0 * y0,
+            `" fill="#${color}" ${!sub ? `transform="translate(${dR}, ${dR})"` : ""} />`
+        ].join(' ');
+    }
+    public renderCurrentCluster(config?: MapMarkerAssetConfig): void {
```
Comment (Ekhorn) [inline ui/component/or-map/src/mapwidget.ts]: This function is very hard to follow and it is not entirely clear how the marker asset config is suppose to behave with clusters. At least add a JSDoc comment describing the configuration options and/or splitting out parts of the code to make it more readable.

```suggestion
    /** */
    public renderCurrentCluster(config?: MapMarkerAssetConfig): void {
```

File: ui/app/manager/src/pages/page-map.ts [hunk 8]
```diff
@@ -340,32 +386,35 @@ export class PageMap extends Page<MapStateKeyed> {
...
-              const attr = asset.attributes[WellknownAttributes.LOCATION] as Attribute<GeoJSONPoint>;
-              return !attr.meta || !attr.meta.hasOwnProperty(WellknownMetaItems.SHOWONDASHBOARD) || !!Util.getMetaValue(WellknownMetaItems.SHOWONDASHBOARD, attr);
-          })
-            .sort((a,b) => {
-                if (a.attributes[WellknownAttributes.LOCATION].value && b.attributes[WellknownAttributes.LOCATION].value){
-                    return b.attributes[WellknownAttributes.LOCATION].value.coordinates[1] - a.attributes[WellknownAttributes.LOCATION].value.coordinates[1];
-                } else {
-                    return;
+            this._assets.filter((asset) => {
+                if (!asset.attributes || assetIdsToShow.findIndex((idx: string) => { return idx === asset.id; }) === -1) {
```
Comment (Ekhorn) [inline ui/app/manager/src/pages/page-map.ts]:
```suggestion
                if (!asset.attributes || assetIdsToShow.findIndex((idx: string) => idx === asset.id) === -1) {
```

File: ui/app/manager/src/pages/page-map.ts [hunk 9]
```diff
@@ -252,6 +281,23 @@ export class PageMap extends Page<MapStateKeyed> {
...
                 this._store.dispatch(setAssets(assets));
 
+                if (this._map) {
+                    this.locationAssets = [];
+                    this.assetTypes = [];
+                    this._map.cleanUpMarker();
+
+                    assets.forEach((asset: Asset) => {
+                        if (this._map.addMarker(asset)) {
+                            if(!this.assetTypes.find((assetType: string) => { return assetType === asset.type; })) {
```
Comment (Ekhorn) [inline ui/app/manager/src/pages/page-map.ts]:
```suggestion
                            if(!this.assetTypes.find((assetType: string) => assetType === asset.type)) {
```

File: ui/component/or-map/src/mapwidget.ts [hunk 10]
```diff
@@ -656,6 +749,90 @@ export class MapWidget {
...
+            if (!marker) { 
+                const placeholder = document.createElement("div");
+                marker = this.cachedMarkers[id] = new maplibregl.Marker({ element: placeholder }).setLngLat(coords);
+            }
+            newMarkers[id] = marker;
+
+            if (!this.markersOnScreen[id]) {
+                marker.addTo(this._mapGl);
+                this.assetsOnScreen[id] = JSON.parse(feature.properties.asset);
+            };
```
Comment (Ekhorn) [inline ui/component/or-map/src/mapwidget.ts:758]: Consider moving `JSON.parse(feature.properties.asset)` upwards to improve performance slightly. Using `feature.geometry` calls a getter which calls [toGeoJSON](https://github.com/maplibre/maplibre-gl-js/blob/d815d2977c19972b589363f44494aaae11ebf593/src/source/vector_tile_mlt.ts#L49) while we already have the coordinates present on the asset itself. (Might be up to a 5% improvement)

```suggestion
            const asset = JSON.parse(feature.properties.asset);
            const point = asset.attributes.location.value.coordinates;

            let marker = this.cachedMarkers[id]
            if (!marker) { 
                const placeholder = document.createElement("div");
                marker = this.cachedMarkers[id] = new maplibregl.Marker({ element: placeholder }).setLngLat(point.coordinates);
            }
            newMarkers[id] = marker;

            if (!this.markersOnScreen[id]) {
                marker.addTo(this._mapGl);
                this.assetsOnScreen[id] = asset;
            };
```

File: ui/component/or-map/src/mapwidget.ts [hunk 11]
```diff
@@ -656,6 +749,90 @@ export class MapWidget {
...
+        for (const feature of features) {
+            if (!feature.properties.id) continue;
+            const id: string = feature.properties.id;
+            const geometry = feature.geometry as Geometry & { coordinates: LngLatLike };
+            const coords = geometry.coordinates;
+
+            let marker = this.cachedMarkers[id]
+            if (!marker) { 
+                const placeholder = document.createElement("div");
+                marker = this.cachedMarkers[id] = new maplibregl.Marker({ element: placeholder }).setLngLat(coords);
```
Comment (Ekhorn) [inline ui/component/or-map/src/mapwidget.ts]: Already checked this I think, but if possible avoid creating these markers if we can.

File: ui/component/or-map/src/or-map-legend.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+                    icon: icon,
+                    color: color,
+                    label: label,
+                };
+            });
+        }
+
+        if (this._showLegend) {
+            this._showLegend.hidden = true;
+        }
```
Comment (Ekhorn) [inline ui/component/or-map/src/or-map-legend.ts]: Move up into `assetTypes` changedProperties? To avoid it collapsing after re-render?

File: ui/component/or-map/src/mapwidget.ts (hunk 10 above)
Comment (Ekhorn) [reply ui/component/or-map/src/mapwidget.ts:758]: Unsure about this won't apply

File: ui/component/or-map/src/mapwidget.ts [hunk 13]
```diff
@@ -467,20 +500,80 @@ export class MapWidget {
...
+                    'circle-radius': 0,
+                    // 'circle-color': '#11b4da',
+                    // 'circle-radius': 4,
+                    // 'circle-stroke-width': 1,
+                    // 'circle-stroke-color': '#fff'
+                }
+            });
+        }
+
+        this._mapGl.on('zoom', () => {
```
Comment (Ekhorn) [inline ui/component/or-map/src/mapwidget.ts]: Add or remove click behavior from https://maplibre.org/maplibre-gl-js/docs/examples/create-and-style-clusters/

File: ui/component/model/tsconfig.json [hunk 14]
```diff
@@ -3,7 +3,8 @@
   "compilerOptions": {
     "outDir": "./lib",
     "rootDir": "src",
-    "preserveConstEnums": true
+    "preserveConstEnums": true,
+    "skipLibCheck": true
```
Comment (MartinaeyNL) [inline ui/component/model/tsconfig.json]: What is the reason for this `skipLibCheck` change?

File: ui/app/manager/src/pages/page-map.ts
Comment (MartinaeyNL) [inline ui/app/manager/src/pages/page-map.ts:1]: I would adjust the CSS for mobile screens.
Enabling the legend almost covers the full screen with the available asset types, which is annoying.
We should also make the legend collapsed by default when the map on a small screen.

Image

File: ui/component/or-map/src/index.ts [hunk 15]
```diff
@@ -1,32 +1,39 @@
...
 // Re-exports
-export {Util, LngLatLike, LngLat};
+export {Util, LngLatLike, LngLat, ClusterConfig};
 export * from "./markers/or-map-marker";
 export * from "./markers/or-map-marker-asset";
 export {IControl} from "maplibre-gl";
 export * from "./or-map-asset-card";
+export * from "./or-map-legend";
+
+export interface LocationAsset extends Asset {
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/index.ts]: I would rename this to `AssetWithLocation` or something similar.
The name `LocationAsset` could be confusing since it is not a custom asset type of some sort.

File: ui/app/manager/src/pages/page-map.ts [hunk 16]
```diff
@@ -340,33 +373,27 @@ export class PageMap extends Page<MapStateKeyed> {
...
-                if (a.attributes[WellknownAttributes.LOCATION].value && b.attributes[WellknownAttributes.LOCATION].value){
-                    return b.attributes[WellknownAttributes.LOCATION].value.coordinates[1] - a.attributes[WellknownAttributes.LOCATION].value.coordinates[1];
-                } else {
-                    return;
+
+            ${this.config?.legend?.show && this.assetTypes.length > 1 ? html`<or-map-legend .assetTypes="${this.assetTypes}" @or-map-legend-changed="${(e: OrMapLegendEvent) => {
+                if (this._map) {
+                    this.exclude = e.detail;
+                    this.updateMarkers();
+                    this._map.reload();
```
Comment (MartinaeyNL) [inline ui/app/manager/src/pages/page-map.ts]: I would move this to a separate callback function, instead of this inline approach.

File: ui/app/manager/src/pages/page-map.ts [hunk 17]
```diff
@@ -181,6 +206,10 @@ export class PageMap extends Page<MapStateKeyed> {
     protected assetSubscriptionId: string;
     protected attributeSubscriptionId: string;
 
+    protected assetsOnScreen: LocationAsset[] = [];
+    protected assetTypes: string[] = [];
+    protected exclude: string[] = [];
```
Comment (MartinaeyNL) [inline ui/app/manager/src/pages/page-map.ts]: Refactor these to `_variableName` for consistency 😉 

I think we can also improve the names of these variables.
For example `exclude` should become `excludedTypes` of some sort.
I'm also not a big fan of the `assetsOnScreen` naming.

File: ui/app/manager/src/pages/page-map.ts [hunk 18]
```diff
@@ -252,6 +281,10 @@ export class PageMap extends Page<MapStateKeyed> {
 
                 this._store.dispatch(setAssets(assets));
 
+                if (this._map) {
+                    this.updateMarkers();
+                }
```
Comment (MartinaeyNL) [inline ui/app/manager/src/pages/page-map.ts]: Shouldn't this ` if(this._map) { ` condition be inside the `updateMarkers()` function?

(my single responsibility code guidelines triggered this 😂)

File: ui/app/manager/src/pages/page-map.ts [hunk 19]
```diff
@@ -340,33 +373,27 @@ export class PageMap extends Page<MapStateKeyed> {
...
-            .map(asset => {
-                return html`
-                            <or-map-marker-asset ?active="${this._currentAsset && this._currentAsset.id === asset.id}" .asset="${asset}" .config="${this.config.markers}"></or-map-marker-asset>
-                        `;
-            })
-        }
+            }}"></or-map-legend>` : null}
+
+            <or-map id="map" class="or-map" .cluster="${this.config.clustering}" showGeoCodingControl @or-map-geocoder-change="${(ev: OrMapGeocoderChangeEvent) => {this._setCenter(ev.detail.geocode);}}">
+                ${this.assetsOnScreen.sort((a,b) => {;
```
Comment (MartinaeyNL) [inline ui/app/manager/src/pages/page-map.ts]:
```suggestion
                ${this.assetsOnScreen.sort((a,b) => {
```

File: ui/component/or-map/src/markers/or-cluster-marker.ts [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+                w
+            }" height="${
+                w
+            }" viewbox="0 0 ${
+                w
+            } ${
+                w
+            }" text-anchor="middle" style="font: ${
+                fontSize
+            }px Helvetica Neue,Arial,Helvetica,sans-serif; display: block">`;
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/markers/or-cluster-marker.ts]: The formatting looks really weird on this, can we fix it? 😂

File: ui/component/or-map/src/markers/or-cluster-marker.ts [hunk 21]
```diff
@@ -0,0 +1,129 @@
+import { html, LitElement } from "lit";
+import { customElement } from "lit/decorators.js";
+
+type Slice = [name: string, color: string, count: number]
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/markers/or-cluster-marker.ts]: Shouldn't we make this a proper interface that is exported as well?
Or is this something copied over from maplibregl?
If so, we should add docs to this.

File: ui/component/or-map/src/markers/or-cluster-marker.ts
Comment (MartinaeyNL) [inline ui/component/or-map/src/markers/or-cluster-marker.ts:1]: This file overall looks really messy to me.
Can we make it more professional? And if possible, reuse any code from the maplibre dependency?
Would be pity if have to add the "copied over from X" comment everywhere.

File: ui/component/or-map/src/markers/or-map-marker.ts [hunk 22]
```diff
@@ -374,4 +374,12 @@ export class OrMapMarker extends LitElement {
             && this.lat >= -90 && this.lat < 90
             && this.lng >= -180 && this.lng < 180;
     }
+
+    disconnectedCallback(): void {
+        super.disconnectedCallback();
+        // Safely cleanup markers once removed from the DOM
+        if (this._actualMarkerElement?.parentElement) {
+            this._actualMarkerElement.parentElement.removeChild(this._actualMarkerElement);
+        }
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/markers/or-map-marker.ts]: Doesn't this work?
```suggestion
        this._actualMarkerElement?.parentElement?.removeChild(this._actualMarkerElement);
```

File: ui/component/or-map/src/util/index.ts
Comment (MartinaeyNL) [inline ui/component/or-map/src/util/index.ts:1]: As a nice positive note; I like these new functions in this util class 👍
It makes the code in all the other classes much nicer.

File: ui/component/or-map/src/util/mapbox-url.ts
Comment (MartinaeyNL) [inline ui/component/or-map/src/util/mapbox-url.ts:1]: I don't know if this is relevant;

As we move this file (and the util file) into a new `/util` directory, the export path changes.
So, in this case, the import would be `import ... from "@openremote/or-map/util/mapbox-url"`.
Does this change things for custom-projects, or other places as it introduces a breaking change?

File: ui/component/or-map/src/index.ts [hunk 23]
```diff
@@ -397,6 +309,9 @@ export class OrMap extends LitElement {
     @property({type: String})
     public type: MapType = manager.mapType;
 
+    @property({type: String})
+    public cluster?: ClusterConfig;
+
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/index.ts]: Shouldn't this be `@property({type: Object})` ?

File: ui/component/or-map/src/mapwidget.ts [hunk 24]
```diff
@@ -62,15 +73,26 @@ export class MapWidget {
...
+    protected _clusterConfig?: ClusterConfig;
+    protected _pointsMap: any = {
+        type: "FeatureCollection",
+        features: []
+    };
+
+    protected _assetTypesColors: any = {};
+    protected cachedMarkers: Record<string, maplibregl.Marker> = {};
+    protected markersOnScreen: Record<string, maplibregl.Marker> = {};
+    protected assetsOnScreen: Record<string, LocationAsset> = {};
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/mapwidget.ts]: I would add the `_` prefix on these protected fields here as well.
Preferably I'd also see some documentation, but the variable names are clear enough, so it's OK. 👍

File: ui/component/or-map/src/mapwidget.ts [hunk 25]
```diff
@@ -467,20 +497,73 @@ export class MapWidget {
...
+            this._mapGl.on("load", () => this.load())
         }
 
         this._mapContainer.dispatchEvent(new OrMapLoadedEvent());
         this._loaded = true;
         this.createBoundaryBox()
     }
 
+    public load() {
+        if (!this._mapGl) return;
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/mapwidget.ts]: I would add a `console.log()` or `.warn()` of some sort.
It prevents annoying debug sessions haha

File: ui/component/or-map/src/mapwidget.ts
Comment (MartinaeyNL) [inline ui/component/or-map/src/mapwidget.ts:1]: I would add documentation to the `.load()`, `.build()` and all other functions in this class.
Just to explain in short what the function **actually** does.

File: ui/component/or-map/src/or-map-legend.ts [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+
+    protected assetTypesInfo: any;
+
+    protected excluded: string[] = [];
+
+    @query("#legend-content")
+    protected _showLegend?: HTMLDivElement;
+
+    protected shouldUpdate(_changedProperties: PropertyValues): boolean {
+        if (_changedProperties.has("assetTypes")) {
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/or-map-legend.ts]: You can remove the underscore inside the parameter, as it is not unused anymore 😉

File: ui/component/or-map/src/or-map-legend.ts [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+        if (this._showLegend) {
+            this._showLegend.hidden = !this._showLegend?.hidden;
+        }
+    }
+
+    protected render() {
+        return html`
+            <div id="legend">
+                <div id="legend-title" @click="${(evt: MouseEvent) => this._onHeaderClick(evt)}">
+                    <span>${i18next.t("mapPage.legendTitle")}</span><or-icon style="cursor: pointer" icon="menu"></or-icon>
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/or-map-legend.ts]: Replace the `i18next` span element with `<or-translate>`

File: ui/component/or-map/src/or-map-legend.ts [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+@customElement("or-map-legend")
+export class OrMapLegend extends LitElement {
+    @property({ type: Array })
+    public assetTypes: string[] = [];
+
+    protected assetTypesInfo: any;
+
+    protected excluded: string[] = [];
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/or-map-legend.ts]: I assume this should be `excludedTypes: string[]` ?

File: ui/component/or-map/src/or-map-legend.ts [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+        return super.shouldUpdate(_changedProperties);
+    }
+
+    static get styles(): CSSResultGroup {
+        return mapAssetLegendStyle;
+    }
+
+    protected _onHeaderClick(evt: MouseEvent | null) {
+        if (this._showLegend) {
+            this._showLegend.hidden = !this._showLegend?.hidden;
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/or-map-legend.ts]: You can remove the `?` here.

File: ui/component/or-map/src/or-map-legend.ts [hunk 30]
```diff
@@ -0,0 +1,105 @@
...
+                                    icon="${this.assetTypesInfo[assetType].icon}"
+                                    style="color: #${this.assetTypesInfo[assetType].color}"
+                                ></or-icon>
+                                <span id="asset-label" style="flex: 1">${this.assetTypesInfo[assetType].label}</span>
+                                <or-mwc-input
+                                    .type="${InputType.CHECKBOX}"
+                                    .value="${!this.excluded.includes(assetType)}"
+                                    @or-mwc-input-changed="${(ev: OrInputChangedEvent) => {
+                                        if (ev.detail.value) {
+                                            this.excluded.splice(this.excluded.indexOf(assetType), 1);
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/or-map-legend.ts]: Does `.splice()` update the state correctly?
Or should it explicitly **not** call a render lifecycle for the `or-map-legend` ?

File: ui/component/or-map/src/style.ts [hunk 31]
```diff
@@ -241,4 +241,73 @@ export const mapAssetCardStyle = css`
...
+            }
+    
+            #legend-content {
+                -webkit-box-shadow: 1px 1px 2px 0px rgba(0,0,0,0.28);
+                -moz-box-shadow: 1px 1px 2px 0px rgba(0,0,0,0.28);
+                box-shadow: 1px 1px 2px 0px rgba(0,0,0,0.28);
+                background-color: var(--internal-or-map-asset-legend-background-color);
+                margin-top: 10px;
+                border-radius: 3px;
+                padding: 10px;
```
Comment (MartinaeyNL) [inline ui/component/or-map/src/style.ts]:
```suggestion
                padding: 5px 0 5px 10px;
```

File: ui/component/rest/tsconfig.json [hunk 32]
```diff
@@ -2,7 +2,8 @@
   "extends": "../../tsconfig",
   "compilerOptions": {
     "outDir": "./lib",
-    "rootDir": "src"
+    "rootDir": "src",
+    "skipLibCheck": true
```
Comment (MartinaeyNL) [inline ui/component/rest/tsconfig.json]: Any reason for the introduction of `skipLibCheck` ?

File: .yarnrc.yml [hunk 33]
```diff
@@ -4,3 +4,5 @@ yarnPath: .yarn/releases/yarn-4.10.3.cjs
 npmMinimalAgeGate: 10080 # A week in minutes
 npmPreapprovedPackages:
   - "@openremote/*"
+  - "@maplibre/*"
+  - "maplibre-gl"
```
Comment (MartinaeyNL) [inline .yarnrc.yml]: I wouldn't add these to the preapproved packages.
It's OK for local testing, but remove these before merging.

File: ui/component/or-map/src/or-map-legend.ts (hunk 30 above)
Comment (Ekhorn) [reply ui/component/or-map/src/or-map-legend.ts]: Functionally it works. Adding `@state` doesn't seem to cause a re-render, but that shouldn't be needed right?

File: ui/component/or-map/src/util/mapbox-url.ts
Comment (Ekhorn) [reply ui/component/or-map/src/util/mapbox-url.ts:1]: Nope only internal usage AFAIK

File: ui/component/or-map/src/util/index.ts
Comment (Ekhorn) [reply ui/component/or-map/src/util/index.ts:1]: Thx! Ye there is a lot more we can probably move, but want to leave that until we start redoing the entire thing :wink:

File: ui/component/or-map/src/markers/or-cluster-marker.ts
Comment (Ekhorn) [reply ui/component/or-map/src/markers/or-cluster-marker.ts:1]: Changed the formatting a bit..

File: ui/component/or-map/src/mapwidget.ts
Comment (Ekhorn) [reply ui/component/or-map/src/mapwidget.ts:1]: Added short descriptions to the methods mentioned. We can describe stuff when we actually rework everything. Want to avoid adding descriptions for methods we won't be needing after the rework.

File: ui/app/manager/src/pages/page-map.ts
Comment (Ekhorn) [inline ui/app/manager/src/pages/page-map.ts]: Update markers  properly when realm changed


```suggestion
if (this._map) {
   this._updateMarkers();
   this._map.refresh();
   this._map.reload();
}
```

File: ui/component/or-map/src/markers/or-map-marker-asset.ts
Comment (Ekhorn) [inline ui/component/or-map/src/markers/or-map-marker-asset.ts:1]: Broken styling for marker direction decoration.

File: ui/component/or-map/src/index.ts
Comment (Ekhorn) [inline ui/component/or-map/src/index.ts:1]: The markers aren't rendered anymore on the assets page.

File: ui/component/or-map/src/mapwidget.ts
Comment (Ekhorn) [inline ui/component/or-map/src/mapwidget.ts:1]: Some marker cluster appear to contain multiple asset types while only 1 asset type is selected.

File: ui/app/manager/test/map.test.ts
Comment (Ekhorn) [inline ui/app/manager/test/map.test.ts:1]: Write tests for:

- [x] Marker config
- [x] Realm switching with toggled types
- [x] Direction
- [x] WSocket updates
- [x] Show on dashboard

File: ui/app/manager/src/pages/page-map.ts [hunk 34]
```diff
@@ -300,11 +335,14 @@ export class PageMap extends Page<MapStateKeyed> {
               this._assets = [];
           }
           this.unsubscribeAssets();
-          this.subscribeAssets(realm);
-
-          if (this._map) {
-              this._map.refresh();
-          }
+          this.subscribeAssets(realm).then(() => {
+              console.log(realm)
```
Comment (MartinaeyNL) [inline ui/app/manager/src/pages/page-map.ts]: Accidental `console.log` 😉

File: ui/app/manager/src/pages/page-map.ts [hunk 35]
```diff
@@ -300,11 +335,14 @@ export class PageMap extends Page<MapStateKeyed> {
...
+              if (this._prevRealm && this._prevRealm !== realm) {
```
Comment (MartinaeyNL) [inline ui/app/manager/src/pages/page-map.ts]: Is this `_prevRealm` check necessary? I don't think so, but I'm not 100% sure...

File: ui/app/manager/test/map.test.ts [hunk 36]
```diff
@@ -1,35 +1,224 @@
 import { expect } from "@openremote/test";
-import { test, userStatePath } from "./fixtures/manager.js";
-import { preparedAssetsWithLocation as assets } from "./fixtures/data/assets.js";
+import { adminStatePath, test, userStatePath } from "./fixtures/manager.js";
+import { preparedAssetsWithLocation as assets, assignLocation, BBox, getAssetTypeColour, getAssetTypes, randomAsset } from "./fixtures/data/assets.js";
+import { OrMap } from "@openremote/or-map/src/index.js";
```
Comment (MartinaeyNL) [inline ui/app/manager/test/map.test.ts]: Is this the correct import path?

File: ui/app/manager/test/map.test.ts [hunk 37]
```diff
@@ -1,35 +1,224 @@
...
+import { Asset, AssetTypeInfo } from "@openremote/model";
+import { OrClusterMarker } from "@openremote/or-map/lib/markers/or-cluster-marker.js";
```
Comment (MartinaeyNL) [inline ui/app/manager/test/map.test.ts]: Is this the correct import path?

File: ui/app/manager/test/fixtures/data/assets.ts [hunk 38]
```diff
@@ -126,3 +126,62 @@ export const agent: Asset = {
...
+  const validAssetInfos = assetInfos
+    .filter(({ assetDescriptor }) => assetDescriptor?.name && assetDescriptor.descriptorType === "asset");
+
+  const randomIndex = Math.round(randomBetween(0, validAssetInfos.length - 1));
+  const info = validAssetInfos[randomIndex];
+
+  const type = info.assetDescriptor!.name;
+  const attributes = Object.fromEntries(
+    Object.values(info?.attributeDescriptors ?? {}).map(({ name, type }) => ([name, { name, type, meta: {} }]))
+  );
```
Comment (MartinaeyNL) [inline ui/app/manager/test/fixtures/data/assets.ts]: Why do we use a "randomly picked asset out of a list" inside tests?
The word 'random' does not sound consistent haha
I personally tend to hardcode whatever number in the tests code.

File: ui/app/manager/test/map.test.ts [hunk 39]
```diff
@@ -1,35 +1,224 @@
...
+   * @then Cluster marker disappears
+   * @and All 10 asset markers part of the cluster are shown
+   *
+   * @skip This test is skipped on Firefox because headless mode does not support WebGL required by maplibre
+   */
+  test("should display clustered markers", async ({ page, manager, browserName }) => {
+    test.skip(browserName === "firefox", "firefox headless mode does not support webgl required by maplibre");
+
+    const bbox = { west: 4.4857, south: 51.9162, east: 4.4865, north: 51.9167 };
+    const assetInfos = (await manager.axios.request<AssetTypeInfo[]>({ url: "/model/assetInfos" })).data;
```
Comment (MartinaeyNL) [inline ui/app/manager/test/map.test.ts]: Can't we use `manager.rest.api.AssetModelResource.getAssetInfos()` ?

**Comment by Ekhorn** [comment]
Closed in favor of https://github.com/openremote/openremote/pull/2284

File: ui/app/manager/test/fixtures/data/assets.ts (hunk 38 above)
Comment (Ekhorn) [reply ui/app/manager/test/fixtures/data/assets.ts]: Are asset infos always returned in a sorted list?

Comment (MartinaeyNL) [reply ui/app/manager/test/fixtures/data/assets.ts]: Based on the Java code, it looks like it.
May not be alphabetical, but I think the order is consistent.
I tend to just get the first or last entry, so the consistency doesn't matter at that point.
Because I wouldn't hardcode the name of the 1st entry for example.

See `ValueUtil.java`:
```
public static AssetTypeInfo[] getAssetInfos(String parentType) {
    return assetInfoMap.values().toArray(new AssetTypeInfo[0]);
}
```
